File: 2. Array and String/pair_product/pair_product.js
// const pairProduct = (numbers, targetProduct) => {
//   let map = {};
//   for(let index=0;index<numbers.length;index++){
    
//     if(!map[numbers[index]])
//       map[numbers[index]] = index;
    
//     if(targetProduct % numbers[index] === 0){
//       let complement = targetProduct/numbers[index];

//       if(map[complement])
//         return [map[complement], index]
//     }
//   }
//   return [null,null];
// };


const pairProduct = (numbers, targetProduct) => {
    let map = {};
    for(let index=0; index<numbers.length; index++){
      let complement = targetProduct/numbers[index];
      if(map[complement]){
        return [map[complement], index]
      }else{
        map[numbers[index]] = index;
      }
    }
}

// Complexity
// n = length of array
// Time: O(n)
// Space: O(n)

File: 2. Array and String/6. pair_product/pair_product.py
from math import floor
def pair_product(numbers, target_product):
  numMap = {}
  for index,num in enumerate(numbers):
    complement = floor(target_product/num)
    if(complement in numMap):
        return (numMap[complement],index)
    numMap[num] = index


print(pair_product([3, 2, 5, 4, 1], 8))
print(pair_product([3, 2, 5, 4, 1], 10))
print(pair_product([4, 7, 9, 2, 5, 1], 5))
print(pair_product([4, 7, 9, 2, 5, 1], 35))
print(pair_product([3, 2, 5, 4, 1], 10))
print(pair_product([4, 6, 8, 2], 16))